pkg/validate/validate.go

```diff
@@ -2,39 +2,164 @@ package validate
 
 import (
 	"context"
+	"fmt"
 	"github.com/armory/spinnaker-operator/pkg/apis/spinnaker/v1alpha1"
 	"github.com/armory/spinnaker-operator/pkg/halconfig"
+	"github.com/go-logr/logr"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/webhook/admission/types"
+	"sort"
 )
 
-// Transformers tracks the list of transformers
-var Validators []SpinnakerValidator
+// generators is used to register all ValidatorGenerator
+var generators []ValidatorGenerator
 
 func init() {
-	Validators = append(Validators, &singleNamespaceValidator{})
+	generators = append(generators, &singleNamespaceValidatorGenerator{}, &kubernetesAccountValidatorGenerator{})
 }
 
 type SpinnakerValidator interface {
-	Validate(svc v1alpha1.SpinnakerServiceInterface, hc *halconfig.SpinnakerConfig, options Options) error
+	GetName() string
+	GetPriority() Priority
+	Validate() ValidationResult
+}
+
+// Priority indicates the order in which the validator should be run, or if it doesn't matter (NoPreference).
+// validators with NoPreference = true can be run in parallel
+type Priority struct {
+	NoPreference bool
+	Order        int
+}
+
+type ValidationResult struct {
+	Error error
+	Fatal bool
+}
+
+type ValidatorGenerator interface {
+	Generate(svc v1alpha1.SpinnakerServiceInterface, hc *halconfig.SpinnakerConfig, options Options) ([]SpinnakerValidator, error)
 }
 
 type Options struct {
 	Ctx    context.Context
 	Client client.Client
 	Req    types.Request
+	Log    logr.Logger
+}
+
+type Account interface {
+	GetName() string
+	GetChecksum() string
 }
 
-func Validate(svc v1alpha1.SpinnakerServiceInterface, options Options) error {
+func Validate(svc v1alpha1.SpinnakerServiceInterface, options Options) []error {
 	_, hc, err := v1alpha1.GetConfig(svc, options.Client)
 	if err != nil {
-		return err
+		return []error{err}
+	}
+	validators, err := generateValidators(svc, hc, options)
+	if err != nil {
+		return []error{err}
+	}
+	seq, parallel := splitSequentialAndParallel(validators)
+	results, abort := runInSequence(seq, options.Log)
+	if abort {
+		return collectErrors(results)
+	}
+	results, _ = runInParallel(parallel, options.Log)
+	return collectErrors(results)
+}
+
+func generateValidators(svc v1alpha1.SpinnakerServiceInterface, hc *halconfig.SpinnakerConfig, options Options) ([]SpinnakerValidator, error) {
+	var validators []SpinnakerValidator
+	for _, g := range generators {
+		va, err := g.Generate(svc, hc, options)
+		if err != nil {
+			return nil, err
+		}
+		for _, v := range va {
+			validators = append(validators, v)
+		}
+	}
+	return validators, nil
+}
+
+// splitSequentialAndParallel returns sorted validators to run in sequence, and the ones that can run in parallel
+func splitSequentialAndParallel(validators []SpinnakerValidator) ([]SpinnakerValidator, []SpinnakerValidator) {
```

Comment:
For readability, it would also be nice to compose the validation graph. Something like:

```
validators = []Validator{
   &singleNamespaceValidator{},
   &parallelValidator{
      &kubernetesAccountValidator{},
      ....
  }
}
```

In that case, we don't need the generator but need to pass the options to `SpinnakerValidator.Validate()`. It'd make adding new validators easier IMO.

Reply:
You don't know how many `kubernetesAccountValidator` instances you'll need until runtime (one validator per account), so I think generators are still needed.
But I see the value in having the above kind of priority definition with generators instead of using the `GetPriority` approach. With `GetPriority` though, you can fine tune a specific account to be validated before the others instead of at the same time, but I don't know if that's a real use case. Do you want me to change it?